src/common/cputrace.h

```diff
@@ -0,0 +1,275 @@
+        non_zero_swi_count = 0;
+        zero_swi_count = 0;
+        sum_swi = sum_cyc = sum_cmiss = sum_bmiss = sum_ins = 0;
+    }
+
+    void dump(ceph::Formatter* f, cputrace_flags flags, const std::string& counter = "") const {
+        f->open_object_section("metrics");
+        f->dump_unsigned("sample_count", sample_count);
+        if (flags & HW_PROFILE_SWI) {
+            f->open_object_section("context_switches");
+            f->dump_unsigned("non_zero_count", non_zero_swi_count);
+            f->dump_unsigned("zero_count", zero_swi_count);
+            f->dump_unsigned("total", sum_swi);
+            if (sample_count) {
+                f->dump_float("avg", (double)sum_swi / sample_count);
+            }
+            f->close_section();
+        }
+
+        auto dump_counter = [&](const std::string& name, uint64_t sum) {
+            f->open_object_section(name.c_str());
+            f->dump_unsigned("total", sum);
+            if (sample_count) {
+                f->dump_float("avg", static_cast<double>(sum) / sample_count);
+            }
+            f->close_section();
+        };
+
+        if (flags & HW_PROFILE_CYC && (counter.empty() || counter == "cpu_cycles"))
+            dump_counter("cpu_cycles", sum_cyc);
+
+        if (flags & HW_PROFILE_CMISS && (counter.empty() || counter == "cache_misses"))
+            dump_counter("cache_misses", sum_cmiss);
+
+        if (flags & HW_PROFILE_BMISS && (counter.empty() || counter == "branch_misses"))
+            dump_counter("branch_misses", sum_bmiss);
+
+        if (flags & HW_PROFILE_INS && (counter.empty() || counter == "instructions"))
+            dump_counter("instructions", sum_ins);
+
+        f->close_section();
+    }
+
+    void dump_to_stringstream(std::stringstream& ss, cputrace_flags flags) const {
+        ss << "sample_count: " << sample_count << "\n";
+        if (flags & HW_PROFILE_SWI) {
+            ss << "\ncontext_switches:\n";
+            ss << "    non_zero_count: " << non_zero_swi_count << "\n";
+            ss << "    zero_count: " << zero_swi_count << "\n";
+            ss << "    total: " << sum_swi << "\n";
+            if (sample_count) {
+                ss << "    avg  : " << (double)sum_swi / sample_count << "\n";
+            }
+        }
+
+        auto dump_counter = [&](const std::string& name, uint64_t sum) {
+            ss << name << ":\n";
+            ss << "    total: " << sum << "\n";
+            if (sample_count) {
+                ss << "    avg  : " << (double)sum / sample_count << "\n";
+            }
+        };
+
+        if (flags & HW_PROFILE_CYC)
+            dump_counter("cpu_cycles", sum_cyc);
+
+        if (flags & HW_PROFILE_CMISS)
+            dump_counter("cache_misses", sum_cmiss);
+
+        if (flags & HW_PROFILE_BMISS)
+            dump_counter("branch_misses", sum_bmiss);
+
+        if (flags & HW_PROFILE_INS)
+            dump_counter("instructions", sum_ins);
+    }
+};
+
+struct HW_ctx {
```

Comment:
suggest to initialize all the members - via inplace assignments or dedicated ctor.